src/noit_message_decoder.c

```diff
@@ -48,6 +48,51 @@
     } \
 } while(0)
 
+mtev_boolean
+noit_metric_extract_tags(const char *in, int *inlen,
+                         const char *start_marker, char end_marker,
+                         noit_metric_tagset_builder_t *builder) {
+  /* This function will take a metric name (in) and extract a
+   * <start_marker><taglist><end_marker> from it
+   * and union that with any existing tags.
+   * If successful, it will shorten inlen accordingly.
+   */
+  int st_off = strlen(start_marker);
+  const char *end = in + *inlen - 1;
+  const char *match = strnstrn(start_marker, st_off, in, *inlen);
+  /* we wannt to find the last one */
+  const char *tag_start = match;
+  while(match) {
+    match = strnstrn(start_marker, st_off,
+                     tag_start+1, (*inlen) - (tag_start + 1 - in));
+    if(match) tag_start = match;
+  }
+  if(!tag_start || *end != end_marker) return mtev_false;
+  if(tag_start + st_off == end) { /* empty */
+    *inlen -= st_off + 1;
+    return mtev_true;
+  }
+
+  if(!noit_metric_tagset_builder_add(builder, tag_start + st_off, end - (tag_start + st_off)))
+    return mtev_false;
+
+  *inlen = tag_start - in;
```

Comment:
It looks for the last and because they are differently bracketed it works.
|MT{....}|ST[...] won't match at all because MT{ doesn't match the ending ].
It then cycles pulling at least one off the end until it cannot successfully pull either.